### PR TITLE
Implement resume for previously paused media players

### PR DESCRIPTION
**Pull request by plasmasolutions**
Hi Joshua,

as promised, here is a little help from a fellow Blender dev. I used the local storage to save the previously paused media players and renamed some actions to better reflect what they are doing. I'm not quite happy about the stop-all command (only the name) but as the user can't see it either way, it does not bother me too much.

Hope you like the little addition,
CU at Bcon19?!

### CODE REVIEW

**Comment by Aligorith** [comment]
Hi Thomas,

Thanks for the patch :)

I was wondering why you opted for using local storage vs just storing a global in the background.js and passing that through to the content script? Is it because you can't actually pass state across like that? Admittedly, this is my first web extensions addon, so I'm still getting my head around the pros and cons of different architectures for these ;)

I was under the impression localStorage wasn't such a great option for this as you have to serialise to/from JSON each time you add/remove data from the structure, and it seems more orientated towards cross-session state while playing/paused seems more of an intrasession thing (i.e. only while the browser's open)?

**Comment by plasmasolutions** [comment]
Hi Joshua,

you're welcome! Thanks for making this cool helper!

The reason behind using "Storage" instead of passing around a global sources array is that I 1st wanted to pave the road for later developments and 2nd wanted to have a clear separation between the (volatile) code and the (non-volatile) data. This may come from my work background as I normally deal with huge code bases and databases, but see it beneficial in our context as well: Persistent settings will definitely be saved in local storage once the add-on matures a bit more, and passing around global fields is just valid as long as Firefox does not reload the add-on - which can happen if a developer calls browser.runtime.reload(...) or browser.management.setEnable(...). In short, it's a bad idea (in the browser world) to assume any storage to be non-volatile if it is in memory.

Furthermore, when I paused the media player with F4 and close the browser (& having the tabs automatically stored/restored) - saving data persistently allows me to unpause all media players that I had running before I closed the browser. I considered this a wanted feature, which is why I didn't even considered to design it as a intrasession storage.

Bottom line: Saving it with local storage will give you playback after browser restart and will help you later when you'd like to save add-on settings persistently across sessions - so why not :) ;)

Have a great day, Thomas

**Comment by Aligorith** [comment]
Aha!  Yes, the addon reload/toggle enabled status is indeed something I wasn't aware of. It's a bit like some of the things you have to be aware of with Android app lifecycles.

**Comment by plasmasolutions** [comment]
100% correct - you wouldn't believe how many assumptions (code wise) I see go wrong every day in the Android world ;) And how many I still have to correct on my own side :)

Have a great day!
